### DISCUSSION TITLE
Using TOPAZ for feature recognition on TEM images of Pt replica

### DISCUSSION THREAD

**Discussion by manamana8014** [tbepler/topaz#92] [Q&A]
Dear ladies and gentlemen,

I do quite a lot of EM, but my experience with Linux and programming is almost at the zero, therefore I have some difficulties of adapting TOPAZ for my particular needs. 
I've started with downloading, installing and running the tutorial data/processing - seems to be working, but trying to recapitulate the same with my own data...

I've got a set of 8-bit TIF images from Pt replicas, acquired on Philips CM-10 TEM, equipped with TVIPS TemCam F416. The microscope is old, therefore the only information available is the pixel size value, which corresponds to 1.1 nm/pixel. You'll find in the bottom an exemplary image attached. On that image I've pre-selected with yellow circles for training some number of the features I'd like to detect, according to the phrase in the Topaz Tutorial: "(Before running Topaz, optional): label a small (100-1000, more is likely to give better results) number of particles on your micrographs using the software of your choice"; the red line defines the area on the image, where those features could be detected. The selection was performed in Fiji; the selected ROI's coordinates were saved to the file rois_1_con_001.csv (as GitHub does not support this kind of files I've converted it to TXT), out of which I prepared Particles.txt to be used for training. 
Considering that the images have the pixel value of 1.1 nm/11 Å I've decided to skip the "preprocessing/downsampling" part and moved straight to "Model training" which resulted in the following:

[topaz.txt](https://github.com/tbepler/topaz/files/6177745/topaz.txt)

(I used here a TXT file as the editor has made something weird out of text)

The question is - what did I do wrong/incomplete and what should I do to get the same kind of results you normally get with the tutorial data?
In the future I will need to perform the 2D classification of detected objects to define later some parameters like size, number of elements, area occupancy etc. 

Thanks for help!
Yaroslav

![1_con_001-1](https://user-images.githubusercontent.com/40792183/111908917-594fef80-8a5b-11eb-911e-e990a0c73c9c.jpg)

[rois_1_con_001.txt](https://github.com/tbepler/topaz/files/6177622/rois_1_con_001.txt)
[particles.txt](https://github.com/tbepler/topaz/files/6177623/particles.txt)

**Comment by tbepler** [comment]
Topaz isn't finding the "image_name" header in your particles.txt file for some reason. Based on a quick look at your file, my guess is that the problem is you have extra whitespace after "image_name" (i.e. "image_name ") before the tab column delimiter. If you delete that white space, it should fix the parsing error.

On a related note, you probably don't need to do any downsampling, but I still highly recommend using topaz preprocess to normalize your micrographs. You can normalize without downsampling by not setting the "-s" flag. In your case, I would also set the --affine flag to just do typical normalization.

Let me know if you still have issues with topaz train after fixing the header in your particles.txt file.

**Reply by manamana8014** [reply]
Thanks for the fast response! I'll let to know once I try.

**Reply by manamana8014** [reply]
OK, here is a report of the current progress/issues in my training:

First of all, I have removed extra whitespace after "image_name" in particles.txt file and tried once again on the non-preprocessed images. This resulted in:



\(base) yar@yar-OptiPlex-3020:~$ conda activate topaz
\(topaz) yar@yar-OptiPlex-3020:~$ topaz train -n 100 \
\>              --num-workers=8 \
\>              --train-images data/test/rawdata1/micrographs/ \    
\>              --train-targets data/test/rawdata1/particles.txt \
\>              --save-prefix=saved_models/test/model \
\>              -o saved_models/test/model/model_training.txt
\# Loading model: resnet8
\# Model parameters: units=32, dropout=0.0, bn=on
\# Loading pretrained model: resnet8_u32
\# Receptive field: 71
CudaWarning: module 'torch._C' has no attribute '_cuda_setDevice'
Falling back to CPU.
\# Using device=0 with cuda=False
WARNING: 13 micrographs listed in the coordinates file are missing from the training images. Image names are listed below.
WARNING: missing micrographs are: ['1_con_001', '1_con_002', '1_con_007', '1_controlfix_006', '2_stand5min_005', '2_stand5min_009', '2_stand5min_012', '2_TTXEGTA_003', '2_TTXEGTA_011', '3_stand10min_002', '3_stand10min_013', '4_UnrstandpHwait_003', '5_UnrLowpHFix_005']
\# Loaded 0 training micrographs with 0 labeled particles
ERROR: no training particles specified. Check that micrograph names in the particles file match those in the micrographs file/directory.
Traceback (most recent call last):
  File "/home/yar/miniconda3/envs/topaz/bin/topaz", line 11, in <module>
    load_entry_point('topaz-em==0.2.4', 'console_scripts', 'topaz')()
  File "/home/yar/miniconda3/envs/topaz/lib/python3.8/site-packages/topaz/main.py", line 148, in main
    args.func(args)
  File "/home/yar/miniconda3/envs/topaz/lib/python3.8/site-packages/topaz/commands/train.py", line 632, in main
    load_data(args.train_images,
  File "/home/yar/miniconda3/envs/topaz/lib/python3.8/site-packages/topaz/commands/train.py", line 272, in load_data
    raise Exception('No training particles.')
Exception: No training particles.
(topaz) yar@yar-OptiPlex-3020:~$



_**Thus I realized that Topaz does not work with TIF files but MRC only. 
I've decided to pre-process my TIF files simultaneously scaling these down 2 times to speed up the next steps a bit and have gotten a set of MRC files, which, for some reason, all turned to be vertically flipped (I checked the files in Fiji after they were generated)... Therefore, I had to perform the vertical flip to my files first in Fiji, then pre-process those and finally, I've got a set of the images I could use. Here is the log for that:**_



\(topaz) yar@yar-OptiPlex-3020:~$ topaz convert -s 2 -o data/test/processed/particles.txt data/test/rawdata/particles.txt
\(topaz) yar@yar-OptiPlex-3020:~$ topaz preprocess -v -s 2 -o data/test/processed/micrographs/ data/test/rawdata/2/*.tif
\# processed: 1_con_001
\# processed: 1_con_002
\# processed: 1_con_007
\# processed: 1_controlfix_006
\# processed: 2_stand5min_005
\# processed: 2_stand5min_009
\# processed: 2_stand5min_012
\# processed: 2_TTXEGTA_003
\# processed: 2_TTXEGTA_011
\# processed: 3_stand10min_002
\# processed: 3_stand10min_013
\# processed: 4_UnrstandpHwait_003
\# processed: 5_UnrLowpHFix_005
\(topaz) yar@yar-OptiPlex-3020:~$


_**Training:**_


\(topaz) yar@yar-OptiPlex-3020:~$ topaz train -n 400 \
\>             --num-workers=8 \
\>             --train-images data/EMPIAR-10025/processed/micrographs/ \
\>             --train-targets data/EMPIAR-10025/processed/particles.txt \
\>             --save-prefix=saved_models/EMPIAR-10025/model \
\>             -o saved_models/EMPIAR-10025/model_training.txt
\# Loading model: resnet8
\# Model parameters: units=32, dropout=0.0, bn=on
\# Loading pretrained model: resnet8_u32
\# Receptive field: 71
CudaWarning: module 'torch._C' has no attribute '_cuda_setDevice'
Falling back to CPU.
\# Using device=0 with cuda=False
\# Loaded 13 training micrographs with 1491 labeled particles
\# source	split	p_observed	num_positive_regions	total_regions
\# 0	train	0.000785	42809	54525952
\# Specified expected number of particle per micrograph = 400.0
\# With radius = 3
\# Setting pi = 0.002765655517578125
\# minibatch_size=256, epoch_size=1000, num_epochs=10
\# Done!

[model_training_1st.txt](https://github.com/tbepler/topaz/files/6204567/model_training_1st.txt)



_**Unfortunately, by copying and pasting the commands, I did not notice, that the expected number of the particles was set to 400, which is way too much (on average it's in a range 80-250), so possibly that was the reason for the following result when I tried to extract particle coordinates:**_



\(topaz) yar@yar-OptiPlex-3020:~$ topaz extract -r 14 -x 2 -m saved_models/EMPIAR-10025/model_epoch10.sav \
\> -o data/EMPIAR-10025/topaz/predicted_particles_all_upsampled.txt \
\> data/EMPIAR-10025/processed/micrographs/*.mrc
CudaWarning: module 'torch._C' has no attribute '_cuda_setDevice'
Falling back to CPU.
Killed
\(topaz) yar@yar-OptiPlex-3020:~$ 



_**Thus, right now I repeat training with the number of particles intentionally set only to 100:**_

\(topaz) yar@yar-OptiPlex-3020:~$ topaz train -n 100 \
\>              --num-workers=8 \
\>              --train-images data/EMPIAR-10025/processed/micrographs/ \
\>              --train-targets data/EMPIAR-10025/processed/particles.txt \
\>              --save-prefix=saved_models/EMPIAR-10025/model \
\>              -o saved_models/EMPIAR-10025/model_training.txt
\# Loading model: resnet8
\# Model parameters: units=32, dropout=0.0, bn=on
\# Loading pretrained model: resnet8_u32
\# Receptive field: 71
CudaWarning: module 'torch._C' has no attribute '_cuda_setDevice'
Falling back to CPU.
\# Using device=0 with cuda=False
\# Loaded 13 training micrographs with 1491 labeled particles
\# source	split	p_observed	num_positive_regions	total_regions
\# 0	train	0.000785	42809	54525952
\# Specified expected number of particle per micrograph = 100.0
\# With radius = 3
\# Setting pi = 0.0006914138793945312
WARNING: pi=0.0006914138793945312 but the observed fraction of positives is 0.0007851123809814453 and method is set to GE-binomial.
WARNING: setting method to PN with pi=0.0007851123809814453 instead.
WARNING: if you meant to use GE-binomial, please set pi > 0.0007851123809814453.
\# minibatch_size=256, epoch_size=1000, num_epochs=10

_Done, once again I tried to extract the coordinates and have got the same result:_

\(topaz) yar@yar-OptiPlex-3020:~$ topaz extract -r 14 -x 2 -m saved_models/EMPIAR-10025/model_epoch10.sav \
\> -o data/EMPIAR-10025/topaz/predicted_particles_all_upsampled.txt \
\> data/EMPIAR-10025/processed/micrographs/*.mrc
CudaWarning: module 'torch._C' has no attribute '_cuda_setDevice'
Falling back to CPU.
Killed
(topaz) yar@yar-OptiPlex-3020:~$ 

_**Is it because of a graphic card? (as I have quite an old machine, I'm limited to what's available)

Thanks for the help!**_
